src/Features/Core/Portable/GenerateMember/GenerateConstructor/GenerateConstructorHelpers.cs

```diff
@@ -2,81 +2,87 @@
+                   IsSymbolAccessible(compilation, constructor) &&
+                   IsCompatible(semanticFacts, semanticModel, constructor, expressions);
         }
 
-        private static bool ParameterTypesMatch(SemanticDocument semanticDocument, IList<ITypeSymbol> parameterTypes, IMethodSymbol method)
+        private static bool IsSymbolAccessible(Compilation compilation, ISymbol symbol)
         {
-            if (method == null)
+            if (symbol == null)
+                return false;
+
+            if (symbol is IPropertySymbol { SetMethod: { } setMethod } property &&
+                !IsSymbolAccessible(compilation, setMethod))
             {
                 return false;
             }
 
-            if (parameterTypes.Count != method.Parameters.Length)
+            // Public and protected constructors are accessible.  Internal constructors are
+            // accessible if we have friend access.  We can't call the normal accessibility
+            // checkers since they will think that a protected constructor isn't accessible
+            // (since we don't have the destination type that would have access to them yet).
+            switch (symbol.DeclaredAccessibility)
             {
-                return false;
+                case Accessibility.ProtectedOrInternal:
+                case Accessibility.Protected:
+                case Accessibility.Public:
+                    return true;
+                case Accessibility.ProtectedAndInternal:
+                case Accessibility.Internal:
+                    return compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(symbol.ContainingAssembly);
+
+                default:
+                    return false;
             }
+        }
 
-            var compilation = semanticDocument.SemanticModel.Compilation;
+        private static bool IsCompatible<TExpressionSyntax>(
+            ISemanticFactsService semanticFacts,
+            SemanticModel semanticModel,
+            IMethodSymbol constructor,
+            ImmutableArray<TExpressionSyntax> expressions)
+            where TExpressionSyntax : SyntaxNode
+        {
+            Debug.Assert(constructor.Parameters.Length == expressions.Length);
 
-            for (var i = 0; i < parameterTypes.Count; i++)
+            for (var i = 0; i < constructor.Parameters.Length; i++)
             {
-                var type1 = parameterTypes[i];
-                if (type1 != null)
-                {
-                    var type2 = method.Parameters[i].Type;
-
-                    if (!compilation.HasImplicitConversion(fromType: type1, toType: type2))
-                    {
-                        return false;
-                    }
-                }
+                var constructorParameter = constructor.Parameters[i];
+                var conversion = semanticFacts.ClassifyConversion(semanticModel, expressions[i], constructorParameter.Type);
```

Comment:
Is using semanticFacts instead of the Compilation to classify the conversion the fix for the lambda issue? I can't see where else it would be...

Reply:
generally use.  Compilation can be used to classify between two ITYpeSymbols.  However, SemanticModel (which is what semantic facts uses) can classify between an *arbitrary expression* and an ITypeSymbol.

Consider the case where the expr is `x => x.Length` and the destination type symbol is `Func<string, int>`.  Technically `x => x.Length` has no type of it's own.  So we can't actually classify between it's type and `Func<string, int>`.  *However* we can ask if there's a conversion between that untyped expression node and that type symbol.